### PR TITLE
Show real name

### PR DESCRIPTION
**Pull request by PowerPan**
Add A Config Option To Show the Full Name (First and Last Name) instead of the username

### CODE REVIEW

**Comment by PowerPan** [comment]
@MyIgel @weeman1337 what i have to do for merging?

File: config/config.default.php [hunk 1]
```diff
@@ -261,6 +261,9 @@
 
     // Enables prename and lastname
     'enable_user_name'        => (bool)env('ENABLE_USER_NAME', false),
+    
+    // Enables Showing prename and lastname instead of Username
+    'enable_show_user_name'   => (bool)env('ENABLE_SHOW_USER_NAME', false),
```
Comment (MyIgel) [inline config/config.default.php:266]:
```suggestion
    'enable_show_user_name'   => (bool)env('ENABLE_SHOW_USER_NAME', false) && (bool)env('ENABLE_USER_NAME', false),
```

File: includes/controller/shift_entries_controller.php [hunk 2]
```diff
@@ -117,7 +117,12 @@ function shift_entry_create_controller_admin($shift, $angeltype): array
     $users = User::query()->orderBy('name')->get();
     $users_select = [];
     foreach ($users as $user) {
-        $users_select[$user->id] = $user->name;
+        $name = $user->name;
+        if (config('enable_show_user_name')) {
+            $name = $user->personalData->first_name.' '.$user->personalData->last_name;
```
Comment (MyIgel) [inline includes/controller/shift_entries_controller.php:122]: definitely breaks the linter -> codestyle

File: includes/view/User_view.php [hunk 3]
```diff
@@ -891,13 +893,18 @@ function User_Nick_render($user, $plain = false)
...
+    if (config('enable_show_user_name')) {
+        $name = $user->personalData->first_name.' '.$user->personalData->last_name;
+    }
 
     if ($plain) {
-        return sprintf('%s (%u)', $user->name, $user->id);
+        return sprintf('%s (%u)', $name, $user->id);
     }
-
+
```
Comment (MyIgel) [inline includes/view/User_view.php:905]: Useless indentation

File: includes/view/User_view.php [hunk 4]
```diff
@@ -891,13 +893,18 @@ function User_Nick_render($user, $plain = false)
     if (is_array($user)) {
         $user = (new User())->forceFill($user);
     }
+    
+    $name = $user->name;
+    if (config('enable_show_user_name')) {
+        $name = $user->personalData->first_name.' '.$user->personalData->last_name;
```
Comment (MyIgel) [inline includes/view/User_view.php:899]: Same here: linting

File: resources/views/macros/base.twig [hunk 5]
```diff
@@ -20,7 +20,12 @@
     <a href="{{ opt.url|default(url('users', {'action': 'view', 'user_id': user.id})) }}"
             {%- if not user.state.arrived %} class="text-muted"{% endif -%}
     >
-        {{ _self.angel() }} {{ user.name }}
+        {{ _self.angel() }} 
+        {% set username = user.name %}
+        {% if(config('enable_show_user_name')) %}
+            {% set username = user.personalData.first_name ~ ' ' ~  user.personalData.last_name %}
+        {% endif %}
+        {{ username }}
```
Comment (MyIgel) [inline resources/views/macros/base.twig:28]: Imho that decision should be extracted to its own macro which also can be used in the navbar template

**Comment by weeman1337** [comment]
Thank you for your contribution.

- Can you please have a look at the [CONTRIBUTING](https://github.com/engelsystem/engelsystem/blob/main/CONTRIBUTING.md) and update your code accordingly? `composer run phpcs` can support you here.
- Also we aim for a 100 % auto test code coverage (excluding `/includes`). Can you add some tests?
- If possible can you squash your commits and give them a meaningful message?
- Some feedback on the code of this PR to follow soon.

This features seems to be useful in general. After actioning the items from above we can merge this.

**Review by MyIgel — CHANGES_REQUESTED** [review]
Without further testing and just from a short glance at the code, here are some things that should be changed (in addition to the changes that @weeman1337 mentioned).

**Review by weeman1337 — CHANGES_REQUESTED** [review]
Hey @PowerPan :wave: Can you fix the issues mentioned by previous reviewers and solve the conflicts in your branch? If we don't hear from you for another two weeks, we may unfortunately close this pull request.

**Comment by weeman1337** [comment]
If you want to continue working on this after two weeks, just leave comment here so that we can reopen the pull request.

File: src/Models/User/User.php [hunk 6]
```diff
@@ -261,4 +262,16 @@ public function shiftsUpdated(): HasMany
     {
         return $this->hasMany(Shift::class, 'updated_by');
     }
+
+    public function getDisplayNameAttribute(): string
+    {
+        if (
+            config('display_full_user_name')
+            && !empty($this->personalData->first_name . $this->personalData->last_name)
```
Comment (weeman1337) [inline src/Models/User/User.php:270]: The value is trimmed on output but not for the empty check.

File: tests/Unit/Controllers/MessagesControllerTest.php [hunk 7]
```diff
@@ -57,6 +57,28 @@ public function testIndexUnderNormalConditionsReturnsCorrectViewAndData(): void
...
+    /**
+     * @testdox index: User is shown as first name and last name instead of nickname
+     * @covers \Engelsystem\Controllers\MessagesController::listConversations
+     */
+    public function testIndexUnderNormalConditionsReturnsFormattedUserName(): void
+    {
+        $this->config->set('display_full_user_name', true);
+
+        $this->userA->personalData->first_name = 'Frank';
+        $this->userA->personalData->last_name = 'Nord';
```
Comment (weeman1337) [inline tests/Unit/Controllers/MessagesControllerTest.php:69]:
```suggestion
        $this->userA->personalData->last_name = 'Nerd';
```

File: tests/Unit/Controllers/MessagesControllerTest.php [hunk 8]
```diff
@@ -57,6 +57,28 @@ public function testIndexUnderNormalConditionsReturnsCorrectViewAndData(): void
...
+        $this->userA->personalData->save();
+
+        $this->response->expects($this->once())
+            ->method('withView')
+            ->willReturnCallback(function (string $view, array $data) {
+                $this->assertEquals('Frank Nord', $data['users'][1]);
```
Comment (weeman1337) [inline tests/Unit/Controllers/MessagesControllerTest.php:75]:
```suggestion
                $this->assertEquals('Frank Nerd', $data['users'][1]);
```

File: tests/Unit/Models/User/UserTest.php [hunk 9]
```diff
@@ -480,4 +480,35 @@ public function testShiftsCreatedAndUpdated(): void
...
         $this->assertCount(2, $user2->shiftsUpdated);
     }
+
+    public function getDisplayNameAttributeProvider(): array
+    {
+        return [
+            ['lorem'],
+            ['Foo', 'Foo'],
+            ['Bar', null, 'Bar'],
+            ['Foo Bar', 'Foo', 'Bar'],
```
Comment (weeman1337) [inline tests/Unit/Models/User/UserTest.php:490]: I would add a test case with empty real name and " " real name.

**Review by weeman1337 — CHANGES_REQUESTED** [review]
Looks good :+1:  Some details

File: tests/Unit/Controllers/MessagesControllerTest.php (hunk 8 above)
Comment (MyIgel) [reply tests/Unit/Controllers/MessagesControllerTest.php:75]: That is not an option

File: config/config.default.php [hunk 10]
```diff
@@ -266,9 +266,13 @@
     // Whether the mobile number can be shown to other users
     'enable_mobile_show'      => (bool) env('ENABLE_MOBILE_SHOW', false),
 
-    // Enables prename and lastname
+    // Enables firstname and lastname
```
Comment (weeman1337) [inline config/config.default.php:269]:
```suggestion
    // Enables first name and last name
```

File: config/config.default.php [hunk 11]
```diff
@@ -266,9 +266,13 @@
...
     'enable_user_name'        => (bool) env('ENABLE_USER_NAME', false),
 
+    // Show a users firstname and lastname instead of username
```
Comment (weeman1337) [inline config/config.default.php:272]:
```suggestion
    // Show a user's first name and last name instead of user name
```

File: config/config.default.php [hunk 12]
```diff
@@ -266,9 +266,13 @@
...
+    'display_full_user_name'  => env('DISPLAY_FULL_USER_NAME', false)
```
Comment (weeman1337) [inline config/config.default.php:273]: The description says „instead of user name“ but the option is `display_full_user_name`.

For less confusion I would name this `display_full_name`.

File: src/Models/User/User.php [hunk 13]
```diff
@@ -263,4 +264,16 @@ public function shiftsUpdated(): HasMany
...
         return $this->hasMany(Shift::class, 'updated_by');
     }
+
+    public function getDisplayNameAttribute(): string
+    {
+        if (
+            config('display_full_user_name')
+            && !empty(trim($this->personalData->first_name . $this->personalData->last_name))
+        ) {
+            return trim($this->personalData->first_name . ' ' . $this->personalData->last_name);
```
Comment (weeman1337) [inline src/Models/User/User.php:274]: The names should be trimmed separately here to avoid duplicated white spaces.

```suggestion
            return trim($this->personalData->first_name) . ' ' . trim($this->personalData->last_name);
```

File: tests/Unit/Models/User/UserTest.php [hunk 14]
```diff
@@ -482,4 +482,40 @@ public function testShiftsCreatedAndUpdated(): void
         $this->assertCount(2, $user2->shiftsCreated);
         $this->assertCount(2, $user2->shiftsUpdated);
     }
+
+    public function getDisplayNameAttributeProvider(): array
+    {
+        return [
```
Comment (weeman1337) [inline tests/Unit/Models/User/UserTest.php:488]: Can you add test cases with leading / trailing white spaces here?

**Review by weeman1337 — CHANGES_REQUESTED** [review]
Looks not too bad overall.

I would put the name of the option up for discussion + some smaller nitpicks.

**Review by MyIgel — APPROVED** [review]

**Review by weeman1337 — APPROVED** [review]
Törööö :+1:
